crates/bic/src/main.rs

```diff
@@ -1,26 +1,88 @@
+use anyhow::{Context, Result};
+use argh::FromArgs;
+use bidiff::DiffParams;
 use comde::{Compressor, Decompressor};
 use crossbeam_utils::thread;
-use failure::{err_msg, Fallible};
 use log::*;
 use size::Size;
 use std::{
     fs::{self, File},
     io::{self, Read, Seek, Write},
-    path::Path,
+    path::PathBuf,
     str::FromStr,
     time::Instant,
 };
 
-/// Command-line arguments to bic
-struct Args {
-    free: Vec<String>,
-    partitions: usize,
+/// Generate and apply binary patches
+#[derive(FromArgs, PartialEq, Debug)]
+struct Cli {
+    #[argh(subcommand)]
+    cmd: Command,
+}
+
+#[derive(FromArgs, PartialEq, Debug)]
+#[argh(subcommand)]
+enum Command {
+    Diff(Diff),
+    Patch(Patch),
+    Cycle(Cycle),
+}
+
+/// Write the diff of two files to a patch file
+#[derive(FromArgs, PartialEq, Debug)]
+#[argh(subcommand, name = "diff")]
+struct Diff {
+    #[argh(positional)]
+    older: PathBuf,
+    #[argh(positional)]
+    newer: PathBuf,
+    #[argh(positional)]
+    patch: PathBuf,
+    /// number of partitions
+    #[argh(option, default = "1")]
+    sort_partitions: usize,
+    /// compression method to use
+    #[argh(option, default = "Method::Stored")]
```

Comment:
`argh` looks otherwise pretty nice, but expressions as strings always make me uneasy haha

Reply:
Why `argh` and not `structopt`? Performance differences aside, I'm wary of yet again falling into a Google ecosystem because they 100% of the time go to shit.

Reply:
It's inserted as an expression, and raw expressions weren't supported in attributes till recently I think?

Reply:
@bbqsrc see first sentence of the description, I just wanted to try it. It's smaller than structopt but works quite similar. Converting to structopt is trivial after this refactoring.